src/framework/myCFG.ml

```diff
@@ -181,37 +181,52 @@ let createCFG (file: file) =
   let mkEdges fromNode edges toNode = addCfg' toNode edges fromNode in
   (* Function for finding the next real successor of a statement. CIL tends to
    * put a lot of junk between stuff: *)
-  let realnode is_entry stmt =
+  let realnode (source: stmt option) is_entry stmt =
+    if Messages.tracing then Messages.tracei "cfg" "realnode is_entry=%B stmt=%d\n" is_entry stmt.sid;
     let rec realnode is_entry visited stmt =
+      if Messages.tracing then Messages.trace "cfg" "realnode inner is_entry=%B visited=[%a] stmt=%d: %a\n" is_entry (d_list "; " (fun () x -> Pretty.text (string_of_int x))) visited stmt.sid dn_stmt stmt;
       if List.mem stmt.sid visited then stmt
-      else match (List.hd stmt.succs) with
-        | exception (Failure _) -> if is_entry then stmt else raise Not_found
-        | next -> begin
-            let sid = stmt.sid in
-            match stmt.skind with
-            | Block _ -> realnode is_entry (sid::visited) next
-            | Goto _ -> realnode is_entry (sid::visited) next
-            | Instr [] -> begin
-                if next.sid == stmt.sid
-                then stmt
-                else realnode is_entry (sid::visited) next
-              end
-            | Loop _ -> realnode is_entry (sid::visited) next
-            | If (exp,_,_,_) -> if isZero exp then realnode is_entry (sid::visited) next else stmt
-            | _ -> stmt
+      else
+        let next () = match (List.hd stmt.succs) with
+          | exception (Failure _) -> if is_entry then stmt else raise Not_found
+          | next -> next
+        in
+        let sid = stmt.sid in
+        match stmt.skind with
+        | Block _ -> realnode is_entry (sid::visited) (next ())
+        | Goto _ -> realnode is_entry (sid::visited) (next ())
+        | Instr [] -> begin
+            let next = next () in
+            if next.sid == stmt.sid
```

Comment:
This is probably correct here, but I think we should aim to eliminate usage of physical equality unless this is explicitly what is needed.

Reply:
This was fine because `sid`s are `int`s, but I heavily refactored this function as well and that check wasn't even necessary (the usual cycle detection handles it as well).